src/main/java/net/dries007/tfc/objects/te/TELoom.java

```diff
@@ -0,0 +1,236 @@
+/*
+ * Work under Copyright. Licensed under the EUPL.
+ * See the project README.md and LICENSE.txt for more information.
+ */
+
+package net.dries007.tfc.objects.te;
+
+import javax.annotation.Nonnull;
+import javax.annotation.Nullable;
+
+import net.minecraft.block.Block;
+import net.minecraft.block.state.IBlockState;
+import net.minecraft.entity.player.EntityPlayer;
+import net.minecraft.item.ItemStack;
+import net.minecraft.nbt.NBTTagCompound;
+import net.minecraft.util.EnumHand;
+import net.minecraft.util.ITickable;
+import net.minecraft.util.ResourceLocation;
+import net.minecraftforge.fml.relauncher.Side;
+import net.minecraftforge.fml.relauncher.SideOnly;
+
+import net.dries007.tfc.TerraFirmaCraft;
+import net.dries007.tfc.api.recipes.LoomRecipe;
+import net.dries007.tfc.api.types.Tree;
+import net.dries007.tfc.network.PacketLoomUpdate;
+import net.dries007.tfc.objects.blocks.wood.BlockLoom;
+import net.dries007.tfc.objects.inventory.ingredient.IIngredient;
+
+
+public class TELoom extends TEInventory implements ITickable
+{
+    private Tree cachedWood;
+
+    private int progress = 0;
+
+    private LoomRecipe recipe = null;
+
+    public TELoom()
+    {
+        super(2);
+    }
+
+    @Override
+    public BlockLoom getBlockType()
+    {
+        Block block = super.getBlockType();
+        if (!(block instanceof BlockLoom))
+            throw new IllegalArgumentException("Block type is invalid; must be instance of BlockLoom");
+        return ((BlockLoom) block);
+    }
+
+    @Nullable
+    public Tree getWood()
+    {
+        if (cachedWood == null)
+        {
+            if (world == null)
+                return null;
+            cachedWood = getBlockType().wood;
+        }
+        return cachedWood;
+    }
+
+    @Override
+    public void readFromNBT(NBTTagCompound compound)
+    {
+        super.readFromNBT(compound);
+        progress = compound.getInteger("progress");
+        if (!inventory.getStackInSlot(0).isEmpty())
```

Comment:
This would probably be better off serializing the recipe via registry name. Then you wouldn't have to use any of the output ingredient stuff. which will break as soon as somebody adds multiple recipes that output the same item.

```java
// Deserialization
recipe = nbt.hasKey("recipe") ? TFCRegistries.LOOM.getValue(new ResourceLocation(nbt.getString("recipe"))) : null'
// Serialization
if (recipe != null) nbt.setString("recipe", recipe.getRegistryName().toString());
```